go/vt/vttablet/tabletserver/repltracker/writer_test.go

```diff
@@ -47,16 +48,222 @@ func TestWriteHeartbeat(t *testing.T) {
+// but that it does respond to RequestHeartbeats(), and generates heartbeats on a lease.
+func TestWriteHeartbeatDisabled(t *testing.T) {
+	defaultOnDemandDuration = 3 * time.Second
+
+	db := fakesqldb.New(t)
+	defer db.Close()
+
+	tw := newTestWriter(db, nil, tabletenv.Disable, 0)
+
+	// Even though disabled, the writer will have an on-demand duration set.
+	assert.Equal(t, defaultOnDemandDuration, tw.onDemandDuration)
+
+	db.AddQueryPattern("^INSERT INTO.*", &sqltypes.Result{})
+
+	writes.Reset()
+	writeErrors.Reset()
+
+	tw.writeHeartbeat()
+	lastWrites := writes.Get()
+	assert.EqualValues(t, 1, lastWrites)
+	assert.EqualValues(t, 0, writeErrors.Get())
+
+	t.Run("closed, no heartbeats", func(t *testing.T) {
+		<-time.After(3 * time.Second)
+		assert.EqualValues(t, 1, writes.Get())
+	})
+	{
+		rateLimiter := tw.safeOnDemandRequestsRateLimiter()
+		assert.Nil(t, rateLimiter)
+	}
+	tw.Open()
+	defer tw.Close()
+	{
+		rateLimiter := tw.safeOnDemandRequestsRateLimiter()
+		assert.NotNil(t, rateLimiter)
+	}
+	t.Run("open, no heartbeats", func(t *testing.T) {
+		<-time.After(3 * time.Second)
+		assert.EqualValues(t, 1, writes.Get())
+	})
+	t.Run("request heartbeats, heartbeats", func(t *testing.T) {
+		tw.RequestHeartbeats()
+		ctx, cancel := context.WithTimeout(context.Background(), tw.onDemandDuration-time.Second)
+		defer cancel()
+		ticker := time.NewTicker(1 * time.Second)
+		defer ticker.Stop()
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case <-ticker.C:
+				assert.EqualValues(t, 0, writeErrors.Get())
+				currentWrites := writes.Get()
+				assert.Greater(t, currentWrites, lastWrites)
+				lastWrites = currentWrites
+			}
+		}
+	})
+	t.Run("exhaust lease, no heartbeats", func(t *testing.T) {
+		<-time.After(tw.onDemandDuration)
+		currentWrites := writes.Get()
+		<-time.After(3 * time.Second)
+		assert.EqualValues(t, currentWrites, writes.Get())
+	})
+	tw.Close()
+	{
+		rateLimiter := tw.safeOnDemandRequestsRateLimiter()
+		assert.Nil(t, rateLimiter)
+	}
+}
+
+// TestWriteHeartbeatOnDemand tests that the heartbeat writer initiates leased heartbeats once opened,
+// and then upon RequestHeartbeats().
+func TestWriteHeartbeatOnDemand(t *testing.T) {
+	defaultOnDemandDuration = 7 * time.Second
+	onDemandDuration := 3 * time.Second
+
+	db := fakesqldb.New(t)
+	defer db.Close()
+
+	tw := newTestWriter(db, nil, tabletenv.Heartbeat, onDemandDuration)
+
+	assert.Equal(t, onDemandDuration, tw.onDemandDuration)
+
+	db.AddQueryPattern("^INSERT INTO.*", &sqltypes.Result{})
+
+	writes.Reset()
+	writeErrors.Reset()
+
+	tw.writeHeartbeat()
+	lastWrites := writes.Get()
+	assert.EqualValues(t, 1, lastWrites)
+	assert.EqualValues(t, 0, writeErrors.Get())
+
+	t.Run("closed, no heartbeats", func(t *testing.T) {
+		<-time.After(3 * time.Second)
+		assert.EqualValues(t, 1, writes.Get())
+	})
+	{
+		rateLimiter := tw.safeOnDemandRequestsRateLimiter()
+		assert.Nil(t, rateLimiter)
+	}
+	tw.Open()
+	defer tw.Close()
+	{
```

Comment:
Personal preference, but I don't see the value that these code blocks add and use of them is pretty non-standard within Vitess.

Reply:
I'll explain why I use this: it's allowing me to redeclare a variable (`rateLimiter := ` in this case) that I do not need to fear is "leaking" elsewhere. Also, I dislike how in unit testing you want to duplicate some tests, but `golang` requires that the first assignment be `:=` and then you can only use `=`. But then as you shuffle things, you need to again modify `:=` to `=` or `=` to `:=` in places that are unrelated to your change. This causes git diffs that are irrelevant to your change (not unlike trailing comma in a table definition on an unrelated index when you add a new index).
So I feel like these code blocks help me isolate:
- actual code/variables
- git diffs
It's a personal preference and I recognize this is not standard in Vitess. FWIW I've already spread that around quite a bit.